python_modules/dagster/docs/intro_tutorial/types.rst

```diff
@@ -0,0 +1,233 @@
+            }
+        )
+    )
+    def dataframe_input_schema(config_value):
+        # config_value is dictionary with a single element
+        file_type, file_options = list(config_value.items())[0]
+
+        if file_type == 'csv':
+            path = file_options['path']
+            del file_options['path']
+            return pd.read_csv(path, **file_options)
+        elif file_type == 'parquet':
+            return pd.read_parquet(file_options['path'])
+        elif file_type == 'table':
+            return pd.read_table(file_options['path'])
+        else:
+            raise DagsterInvariantViolationError(
+                'Unsupported file_type {file_type}'.format(
+                    file_type=file_type
+                )
+            )
+
+Any input schema is define by a decorated function with a single argument. The argument is the
+format the input schema takes. In this case it is a `Selector`. Selectors are used when you want
+to be able present several different options to the user but force them to select one. In this case,
+for example, it would not make much sense to allow them to say that a single input should be source
+from a csv and a parquet file: They must choose. (In other type systems this might be called an "input
+union.")
+
+Note our selector provides three keys: `csv`, `parquet`, and `table`. Each of these in turn has
+their own subfield selections. You'll note that these are `Dict` require the user to provide all
+required fields, rather than only one.
+
+The remaining step here is that the user must provide code that takes those validated configurated
+files and produces an object of the expected type. In this case that would be the pandas DataFrame
+type. The user can assume the value abides by the provided schema and that any and all default
+values have been applied.
+
+Selectors are very commonly used for these type of schemas, so we've also provided an alternative
+API that removes some boilerplate.
+
+.. code-block:: py
+
+    @input_selector_schema(
+        Selector(
+            {
+                'csv': define_csv_dict_field(),
+                'parquet': define_path_dict_field(),
+                'table': define_path_dict_field(),
+            }
+        )
+    )
+    def dataframe_input_schema(file_type, file_options):
+        if file_type == 'csv':
+            path = file_options['path']
+            del file_options['path']
+            return pd.read_csv(path, **file_options)
+        elif file_type == 'parquet':
+            return pd.read_parquet(file_options['path'])
+        elif file_type == 'table':
+            return pd.read_table(file_options['path'])
+        else:
+            raise DagsterInvariantViolationError('
+                'Unsupported file_type {file_type}'.format(
+                    file_type=file_type
+                )
+            )
+
+You'll note that we no longer need to manipulate the config_value dictionary. It grabs
```

Comment:
I would move this to after line 121 to change it to something like "Selectors are very commonly used for these type of schemas, so we've also provided an alternative API that removes some boilerplate around manipulating the config_value dictionary. Instead, the user-provided function takes the unpacked key and value of config_value directly, since in the case of a selector, the config_value dictionary has only 1 (key, value) pair."

I think this makes it clearer, so that the user knows what to look for.

Reply:
✔️